java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -181,20 +183,66 @@ private FlowComputation(Set<SymbolicValue> symbolicValues, Predicate<Constraint>
     return flows;
   }
 
-  Stream<ExecutionPath> startPath(ExplodedGraph.Edge edge, PSet<Symbol> trackedSymbols) {
-    return new ExecutionPath(null, PCollections.emptySet(), trackedSymbols, ImmutableList.of(), false).addEdge(edge);
+  Stream<ExecutionPath> startPath(ExplodedGraph.Edge edge, PSet<Symbol> trackedSymbols, SameConstraints sameConstraints) {
+    return new ExecutionPath(null, PCollections.emptySet(), trackedSymbols, sameConstraints, ImmutableList.of(), false).addEdge(edge);
+  }
+
+  private static class SameConstraints {
+    private final Map<Symbol, Boolean> sameConstraintsBySymbol;
+    private final Node node;
+
+    SameConstraints(ExplodedGraph.Node startNode, PSet<Symbol> trackedSymbols, List<Class<? extends Constraint>> domains) {
+      this.node = startNode;
+      this.sameConstraintsBySymbol = new LinkedHashMap<>();
+      trackedSymbols.forEach(symbol -> sameConstraintsBySymbol.put(symbol, domains.stream().allMatch(domain -> sameConstraintWhenSameProgramPoint(node, symbol, domain))));
+    }
+
+    SameConstraints(SameConstraints knownSameConstraints, PSet<Symbol> trackedSymbols, List<Class<? extends Constraint>> domains) {
+      this.node = knownSameConstraints.node;
+      this.sameConstraintsBySymbol = new LinkedHashMap<>(knownSameConstraints.sameConstraintsBySymbol);
+      trackedSymbols.forEach(symbol -> sameConstraintsBySymbol.putIfAbsent(symbol, domains.stream().allMatch(domain -> sameConstraintWhenSameProgramPoint(node, symbol, domain))));
+    }
+
+    private static boolean sameConstraintWhenSameProgramPoint(ExplodedGraph.Node currentNode, Symbol symbol, Class<? extends Constraint> domain) {
+      ProgramState programState = currentNode.programState;
+      SymbolicValue sv = programState.getValue(symbol);
+      if (sv == null) {
+        return false;
+      }
+      Constraint constraint = programState.getConstraint(sv, domain);
+      if (constraint == null) {
+        return false;
+      }
+      Collection<Node> siblingNodes = currentNode.siblings();
+      return siblingNodes.isEmpty() || siblingNodes.stream()
```

Comment:
`allMatch` returns `true` for empty stream, so separate test for emptiness is not necessary